### ISSUE TITLE
If a remote TCP syslog server is down, docker does not start.

### ISSUE THREAD

**Issue by tecnobrat** [docker/docker#21966]
If you try to start docker with a remote tcp syslog, which is currently unavailable, docker will fail to start.

This is especially problematic in a production environment where you are trying to ship logs to a central place. You could have network issues to your syslog server, and this will stop your production apps from running.  I would say that having logs be delayed is optimal to not being able to start your server.

If a node is unresponsive:
`$ docker run --log-opt syslog-address=tcp://google.com:1212 --log-driver=syslog nginx`
* wait about a minute *
`docker: Error response from daemon: Failed to initialize logging driver: dial tcp 216.58.216.142:1212: getsockopt: connection timed out.`

Or if it flat our rejects the connection:
`docker run --log-opt syslog-address=tcp://localhost:1212 --log-driver=syslog nginx`
`docker: Error response from daemon: Failed to initialize logging driver: dial tcp 127.0.0.1:1212: getsockopt: connection refused.`

In both of these cases, your container doesn't launch.  We could use UDP, which would avoid this problem, but also it causes issues if you have intermittent network issues, you will lose logs entirely.

I would suggest that docker still starts, but just caches the logs internally and still attempts to connect and dump its logs, instead of crashing.

**Output of `docker version`:**

```
Client:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Thu Mar 10 21:49:11 2016
 OS/Arch:      darwin/amd64

Server:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Thu Mar 10 21:49:11 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 223
 Running: 15
 Paused: 0
 Stopped: 208
Images: 552
Server Version: 1.10.3
Storage Driver: aufs
 Root Dir: /mnt/sda1/var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 1037
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Plugins:
 Volume: local
 Network: null host bridge
Kernel Version: 4.1.19-boot2docker
Operating System: Boot2Docker 1.10.3 (TCL 6.4.1); master : 625117e - Thu Mar 10 22:09:02 UTC 2016
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 3.858 GiB
Name: default
ID: WIKN:KOFK:L2P5:I5VU:JGGX:UPHB:4PCB:L7FC:PX3N:F42A:XC4N:AEGA
Debug mode (server): true
 File Descriptors: 137
 Goroutines: 215
 System Time: 2016-04-12T20:46:33.827097672Z
 EventsListeners: 1
 Init SHA1:
 Init Path: /usr/local/bin/docker
 Docker Root Dir: /mnt/sda1/var/lib/docker
Username: tecnobrat
Registry: https://index.docker.io/v1/
Labels:
 provider=virtualbox
```

**Comment by sandrokeil** [comment]
It would also be useful to use a file buffer instead and send the logs to the syslog server if it is available again.

**Comment by mlaventure** [comment]
Duplicate of #22129 (although different reason, reasoning is the same).

**Comment by thaJeztah** [comment]
@mlaventure the other way around; this was reported before 😄

**Comment by lacarvalho91** [comment]
I'm also having this problem, is there something in the works for this?

**Comment by programmerq** [comment]
$ docker version
    Client:
     Version:      1.11.1
     API version:  1.23
     Go version:   go1.5.4
     Git commit:   5604cbe
     Built:        Wed Apr 27 00:34:20 2016
     OS/Arch:      darwin/amd64
    
    Server:
     Version:      1.11.1
     API version:  1.23
     Go version:   go1.5.4
     Git commit:   8b63c77
     Built:        Tue May 10 10:39:20 2016
     OS/Arch:      linux/amd64
    
    $ docker run --log-opt syslog-address=tcp://google.com:1212 --log-driver=syslog nginx:alpine
    docker: Error response from daemon: Failed to initialize logging driver: dial tcp 209.141.120.170:1212: getsockopt: connection refused.

1.11 still has similar behavior. it did take longer than the originally described "_about a minute_".

**Comment by michaelwilde** [comment]
It appears any logging driver (including Splunk) has an issue if a TCP connection occurs and the remote node is not available.  It would be preferred for the container to launch and some modicum of retry was available so services were not affected.  I'm not exactly sure what happens to the container if it starts and then the remote log receiver goes down.. i'd hope the container doesn't die.

**Comment by gbolo** [comment]
Any update on this? Need this fixed for tcp+tls

**Comment by michaelwilde** [comment]
...cuz we're at a point now where we are having to tell customers... "well, if your remote TCP endpoint for logging isn't available at the time container is started (for whatever reason), your darn container just won't start".  People do want to use more sophisticated logging *AND* they shouldn't have to rely on syslog UDP (Which isn't that great a multiline logs), so they're left with logging to a filesystem, or using some other logging API's within their app... taking us back to 2005.   We've got to get to a point where the container *will* always start if possible the logging drivers are a reliable option.

**Comment by ionutalexandruisac** [comment]
Having the same issue with containers not being able to start if for example the Splunk logging driver has a connection issue. Any news on when we can expect a fix for this? You would expect a soft fail in such situations instead you get a really hard one.

**Comment by sandrokeil** [comment]
You can use [logspout](https://hub.docker.com/r/deis/logspout/) as a workaround.

**Comment by ionutalexandruisac** [comment]
thank you @sandrokeil for the work-around, unfortunately it won't do any good for me.
Any updates regarding this issue? Anybody from Docker actually looking at it?

**Comment by cpuguy83** [comment]
@ionutalexandruisac We've seen it. Logging is tricky, change something and you piss off half the user base, don't change something and you piss of the other half.

Right now docker tries to ensure you never lose logs, unfortunately in this case that means you can't even start your container... and perhaps worse after your container is started if the remote endpoint goes down for an extended period your container will be blocked on I/O.

Probably need an option to allow for lossy logs that works across all drivers, protocols, etc.

@michaelwilde Docker does not support multiline logging anyway, so syslog UDP would be fine here.

**Comment by lkm** [comment]
@cpuguy83 Not losing logs is good thing, but potential downtime is (I assume) worse for everybody. Couldn't you fall back to file based logging if log host is down and then send the logs when it comes back up? Perhaps a somewhat more difficult request to implement.

The option to essentially drop log events when log host down would also be welcomed!

**Comment by ionutalexandruisac** [comment]
@cpuguy83 - Thanks for your reply.
While I understand your point of view, I think @lkm is quite right: downtime is much worse than losing logs.
I guess it would be fairly simple to implement a feature to "allow container to start/run even if the logging driver is unavailable".
You can leave the default as it is now and that feature should be optional maybe, so that users who want it can actually have it if they want to. We have some critical services running on docker and it's really sad to see them crash and being unable to restart just because something happened to Splunk and is temporary unavailable.

**Comment by nickperry** [comment]
The ideal in my opinion would be to expose 3 options for what to do when a remote logging driver cannot reach its end point.

Block
Spool
Discard

Discard could presumably be passed down to the log driver to implement.

Spool could come later and needs significant discussion and engineering. Could it be that it's split into two options - internal spooling handled by the Docker and external spooling, which is passed down to the log driver?

@michaelwilde @glennblock @outcoldman  I guess the Splunk driver would only be able to support discard, not spool, due to the decision to go with HEC?

https://github.com/docker/docker/issues/16207#issuecomment-139678023 (also note next two comments in that thread).

**Comment by nickperry** [comment]
One work around for Splunk customers who want spooling would be to run the universal forwarder on the Docker host with a TCP:// input and point the Splunk Docker log driver at that.

**Comment by outcoldman** [comment]
I like this discussion. 

@michaelwilde in case if link between Splunk and Host will die while container is running - it will keep running, all logs will go to the log configured for the Docker Daemon as error messages similar to `Failed to send log "{MESSAGE}"`.

We have seen that some of our customers have issues with logging driver, which fails the start when driver cannot connect to the remote host.

In 99% of causes this issue is caused by misconfiguration of logging driver, wrong url or wrong route to the Splunk host. We verify connection intentionally just to inform user that splunk logging driver is configured wrong.

In case of big failure, when driver cannot connect to Splunk for some reason (let's say link between Docker Host and Splunk cluster is down) but customer need to scale out right now - we still want to identify user that logging does not work and customer needs to take an action - the easiest action is to switch to json logging driver. Later when link between docker host and client will be fixed - customer can index json log as well to keep all the logs in one place.

@nickperry can this workflow be applied to you? Do you see any issues with it? Better way.

Btw, installing Splunk Forwarder locally with Docker daemon is still a good solution. It can give you a lot, including retries and all other benefits.

Btw, I am working right now on some improvements for the Splunk Logging Driver, one of them will be `--log-opt splunk-verifyconnection=true|false` (which is Discard logic mentioned by @nickperry). If somebody is interesting they can ping me on DockerCon 2016 in Seattle (Splunk will have a booth and a small presentation) and I can show these improvements.

**Comment by nickperry** [comment]
@outcoldman Thanks a lot for that response.

We've experimented with consuming the json-file logs via universal forwarder but multi-line stuff like java stack traces is causing us problems.

log-opt splunk-verifyconnection=true|false sounds great. What are your other enhancements?

**Comment by glennblock** [comment]
+1 on the concerns here. This has been something we have heard brought up
several times by customers using our driver. The new param @outcoldman is
adding is specifically for this purpose. By default the driver behaves as
it does today, but if you opt-in you can override.

@nickperry have you tried using the Transaction command? It can help
assemble the lines when you search. We don't have a native way to handle
this at ingest time via HEC _yet_.

**Comment by michaelwilde** [comment]
@outcoldman @nickperry @cpuguy83 @lkm While we can do as much as we can at splunk to ensure "things work".. this is less about Splunk, and more about how Docker makes a singular decision that if a logging driver (that uses TCP) fails to connect to its destination host, the container does not start.  One can make the argument "if one cannot log, one should not run".  If thats going to be default, then fine.. but we'd all like to see an option where the container will press on and run itself (even with a logging driver CLI arg)--and of course re-attempt connection to said logging driver's remote host.   

Syslog UDP is nice, but most people--when they can--use TCP for syslog (i think the track-ability of TCP pleases most network admins vs the fire, forget, and hope it arrives method of UDP syslogging) #opinion

**Comment by akvadrako** [comment]
It would be really nice if the log drivers were not an additional point of failure. Ideally,
- containers still run when they can't connect to their TCP endpoint
- the driver continuously tries to reconnect if the connections fails at any point

**Comment by LaurentDumont** [comment]
I've just run into this with the Docker Splunk Driver. If for some reason, the remote Splunk server doesn't answer, the container fails to start. I'm a bit surprised that there isn't a fallback method where it could default to some other logging method instead of just stopping.

**Comment by michaelwilde** [comment]
I think docker has a belief that if a container can't log, a container shouldn't run.  A good part of me agrees with the concept, but in practice there's a good use case for a periodic retry as a fallback and let the container start.

**Comment by lkm** [comment]
@michaelwilde I think the one sane way to run this in production is so follow @nickperry approach and run the universal forwarded docker image on every server, and map it's HEC port (8088) to localhost. We use forwarder management to avoid manually configuring all the forwarders for inputs.

**Comment by LaurentDumont** [comment]
The safest solution would probably to log to a file, and ship everything with the forwarder. I can't see a situation (besides a full disk) where the docker daemon can't write. 

The weird thing is that there is a "none" option for the logging driver. It could be the default fallback option if the primary logging driver fails.

**Comment by outcoldman** [comment]
btw, for the Splunk Logging Driver we also have implemented retry logic, so in case of `--log-driver splunk --log-opt splunk-verifyconnection=false` you will be able to start your container, and in case if Splunk is not available - we will keep trying to send logs. See https://github.com/docker/docker/blob/master/docs/admin/logging/splunk.md#advanced-options for details

**Comment by LaurentDumont** [comment]
Weirdly enough, I get `unknown log opt 'splunk-verify-connection' for splunk log driver` with Docker 1.12.1

**Comment by thaJeztah** [comment]
@LaurentDumont the PR was continued in https://github.com/docker/docker/pull/25786, which is on the 1.13 milestone

**Comment by marcusmyers** [comment]
Any more news on this? We are still experiencing this with 2017.15.0-CE.

**Comment by port22** [comment]
Does not work with
```
Docker version 17.06.0-ce-rc4, build 29fcd5d
```
either:
```
# docker run -d --log-driver=fluentd --log-opt="fluentd-address=localhost:24224" 
--log-opt="fluentd-async-connect"  nginx
8cd56b8003bad9592a8f044f4992e137382bedfbe8c5a4450bbc66f26d8f9850
docker: Error response from daemon: failed to initialize logging driver: 
dial tcp [::1]:24224: getsockopt: connection refused.
```

 in contrast to the docs who say:
```
If container cannot connect to the Fluentd daemon, the container stops immediately unless the fluentd-async-connect option is used.
```

**Comment by vozerov** [comment]
@port22 try to run like this:

`docker run -d --log-driver=fluentd --log-opt="fluentd-address=localhost:24224" --log-opt="fluentd-async-connect=true"  nginx`

worked for me

**Comment by bbergshaven** [comment]
Any updates for the syslog driver as well?

**Comment by bitbrain** [comment]
In Amazon ECS I get the following error when the logging driver is not available for some reason: 
```
CannotStartContainerError: API error (500): Failed to initialize logging driver.
```
This seems to be related somehow.

Any news on this?

**Comment by kinnalru** [comment]
What about syslog driver now?

**Comment by shantanugadgil** [comment]
Same issue, same requirement.

I was hoping the `mode` and `max-buffer-size` variable could be helpful, if the remote syslog server would come online after some time. But no such luck! 😢  😞 

```
# docker --version
Docker version 18.03.0-ce, build 0520e24

# docker run -it --log-driver syslog --log-opt syslog-address=tcp://logs.mydomain.com:514 --log-opt mode=non-blocking --log-opt max-buffer-size=4m alpine ping 127.0.0.1
docker: Error response from daemon: failed to initialize logging driver: dial tcp aaa.bbb.ccc.ddd:514: getsockopt: connection refused.
```

**Comment by abstatic** [comment]
I am also facing the same issue with ECS. If for whatever reason the logging endpoint is down, the container won't even start.

Do we need to put our logging machine under HA too :(

**Comment by deevodavis71** [comment]
Same issue using "syslog" logging driver on Mac OSX Docker version 18.03.1-ce, build 9ee9f40. Kind of makes centralised logging pointless if the containers won't even start if the syslog server isn't available! Have tried the "non-blocking" option but it still fails to start...

**Comment by jortkoopmans** [comment]
@deevodavis71 ; I think you need to recreate the containers, in order to get the new host level logging driver settings propagated. You can check the situation using 'docker inspect' on the stuck containers, under the 'LogConfig' section.

Just ran into this problem as well, docker 18.06ce. Going udp for now.

**Comment by SajalGolas** [comment]
Hi folks,
i am also facing same issue when choose splunk as logging driver for container.
Error Statement : docker: Error response from daemon: failed to initialize logging driver: Options http://localhost:8088/services/collector/event/1.0: dial tcp [::1]:8088: connect: connection refused.

Any suggestions ?

**Comment by hinorashi** [comment]
> @port22 try to run like this:
> 
> `docker run -d --log-driver=fluentd --log-opt="fluentd-address=localhost:24224" --log-opt="fluentd-async-connect=true" nginx`
> 
> worked for me

does it use an in-memory ring buffer as with logging [mode](https://docs.docker.com/config/containers/logging/configure/#configure-the-delivery-mode-of-log-messages-from-container-to-log-driver):
```
mode=non-blocking
```
should we use both `mode=non-blocking` and `fluentd-async-connect=true` in combination or just the latter one is enough

**Comment by AndreyByhalenko** [comment]
We use gelf driver and have the same issue.
Tried to set mode = non-blocking but it's not working.
Tried to set it on daemon.json and inside docker-compose as well.
`{
        "log-driver":"gelf",                                                                                                                                                        
        "log-opts":{                                                                                                                                                                
             "gelf-address": "tcp://10.130.100.11:12201",                                                                                                                           
             "mode": "non-blocking"  
}`

**Comment by arnegroskurth** [comment]
We are having the same problem with the `syslog` logging driver even though we are lucky with this one as we only forward container logs to an agent listening on the loopback-interface of the same host so we do not really require tcp here.

Adding a log-cache (configurable size? ring-buffer?) to the docker-daemon to allow re-establishing tcp connections would be ideal. This would then handle the issue both on container startup and when the connection is lost during normal operation.

**Comment by dodancs** [comment]
Does this only happen with TCP? If I would switch to UDP does the container start even tho the connection to syslog is not yet established?

We are having the same issue, sometimes the syslog container does not start before the other containers, so the former containers don't start at all because of syslog not being available. I mean one option would be to use syslog on the host machine, but not having to use these kind of work-arounds would be nice.

**Comment by Vinnie-** [comment]
2024 and that issue is still alive and kicking?! That's amazing. Will it ever be fixed?

**Comment by YuryHrytsuk** [comment]
> Does this only happen with TCP? If I would switch to UDP does the container start even tho the connection to syslog is not yet established?
> 
> We are having the same issue, sometimes the syslog container does not start before the other containers, so the latter containers don't start at all because of syslog not being available. I mean one option would be to use syslog on the host machine, but not having to use these kind of work-arounds would be nice.

There are also (potential) issues with UDP
* https://github.com/moby/moby/issues/8795
* https://github.com/moby/moby/issues/50766

**Comment by YuryHrytsuk** [comment]
> [@Port22](https://github.com/Port22) try to run like this:
> 
> `docker run -d --log-driver=fluentd --log-opt="fluentd-address=localhost:24224" --log-opt="fluentd-async-connect=true" nginx`
> 
> worked for me

@vozerov does it fix original issue that is "containers do not start if TCP endpoint (used by logger) is down"?

UPD: I checked and it fixes the issue (one can start container even if fluentd endpoint is unavailable) 🚀
